content/blog/posts/o3de-motion-matching-2022-04.md

```diff
@@ -0,0 +1,181 @@
+
+{{< image-width src="/images/blog/motion-matching/script-canvas-graph.png" width="100%" alt="Script Canvas Graph for motion matching character control." >}}
+
+For an interactive game, we can use the gamepad's joystick position as a target to generate a desired future trajectory by creating an exponential curve starting at the current character position going towards the character facing direction, and then bend the curve towards the joystick's target position. This process is called trajectory prediction.
+
+{{< video src="/images/blog/motion-matching/trajectory-prediction.mp4" autoplay="true" loop="true" width="100%" info="Gamepad controlled character using motion matching." >}}
+<br>
+<br>
+
+## Data Analysis & Debugging
+
+While motion matching gets rid of the manual process of setting up locomotion in an animation graph, precisely timing transitions and making sure the blends look good, it also takes away some control. Visualizing data and the state of the algorithm and knowing our data is key for debugging an animation glitch.
+
+We provide several debug visualizations: Features like a joint velocity or the trajectory can be visualized together with the query values as well as the feature costs along with some performance metrics. The feature costs over time can be used to see if a visual glitch corresponds with a high cost at that given point in time. A high cost indicates that the algorithm struggled to find a good matching pose for the query. And this again says that our motion database might lack a motion for the given scenario that can be fixed by retaking or adding an extra animation clip.
+
+We also provide data analysis tools to get some insight about what is missing or where we have duplicated animation data. In the middle image below you can see histograms per feature component showing their value distributions across the motion database. They can provide interesting insights, for example, if the motion database is holding more moving forward animations than it has strafing or backward moving animations, or how many fast vs slow turning animations are in the database. This information can be used to see if there is still a need to record some animations or if some types of animation are over-represented and will lead to ambiguity and decrease the quality of the resulting synthesized animation.
+
+The image on the right shows a scatter plot of the features per frame in the motion database. Our multi-dimensional feature data is projected down to two dimensions using [principal component analysis](https://en.wikipedia.org/wiki/Principal_component_analysis). The density of the clusters and the distribution of the samples overall indicate how hard it is for the search algorithm to find a good matching frame candidate. It shows what areas are well covered and where we might lack some animation data. 
```

Comment:
nit: Avoid starting with "It" in the final sentence as it forces me to think about what we are discussing. Maybe "The plot data .."?

Reply:
Changed to "graph"